File: new/算法/排序/js/quickSort.js
function getIndex(arr, low, high) {
  let left = low;
  let right = high;
  const temp = arr[left];
  while (left < right) {
    // 找到小于temp的，给左边的元素
    while (left < right && arr[right] >= temp) {
      right = right - 1;
    }
    arr[left] = arr[right];
    // 找到大于temp的，给右边的元素
    while (left < right && arr[left] <= temp) {
      left = left + 1;
    }
    arr[right] = arr[left];
  }
  arr[left] = temp;

  // ！！！错误-1：这里忘记返回值了
  return left;
}

function quickSort(arr, low, high) {
  if (low >= high) {
    return;
  }

  const index = getIndex(arr, low, high);
  quickSort(arr, low, index - 1);
  quickSort(arr, index + 1, high);
}

const N = 100;
const arr = [];
for (let i = 0; i < N; i++) {
  arr.push((Math.random() * N) | 0);
}
console.log(arr);
// ！！！错误-2：这里是直接修改内部数组
quickSort(arr, 0, arr.length - 1);
console.log(arr);

File: 算法/排序/lua/quickSort.lua
-- 快速排序
-- 个人理解，喂养模式，
-- 从右至左找到小的喂养low，从左向右找到大的喂养high
local function getIndex(arr, low, high)
  local temp = arr[low]
  while (low < high) do
    -- high找到一个小的给low
    while (low < high and arr[high] >= temp) do high = high - 1 end
    arr[low] = arr[high]

    -- high给了一个小的给low，那么low也要还一个大的给high
    while (low < high and arr[low] <= temp) do low = low + 1 end
    arr[high] = arr[low]
  end

  -- 跳出循环时low和high相等,此时的low或high就是temp的正确索引位置
  -- 由原理部分可以很清楚的知道low位置的值并不是temp,所以需要将temp赋值给arr[low]
  arr[low] = temp
  return low
end

local function quickSort(arr, low, high)
  if (low >= high) then return end
  local index = getIndex(arr, low, high)
  quickSort(arr, low, index - 1)
  quickSort(arr, index + 1, high)
end

-- 生成一个随机数组
local arr = {}
for i = 1, 1000000 do table.insert(arr, math.random(1, 1000)) end

local beginTime = os.time();
quickSort(arr, 1, #arr)
local endTime = os.time();
print(string.format('time consume: %ss ', endTime - beginTime))
-- print(table.unpack(arr))
